src/engine/SCons/Tool/applelink.py

```diff
@@ -40,6 +40,114 @@
 from . import link
 
 
+class AppleLinkInvalidCurrentVersionException(Exception):
+    pass
+
+class AppleLinkInvalidCompatibilityVersionException(Exception):
+    pass
+
+
+def _applelib_versioned_lib_suffix(env, suffix, version):
+    """For suffix='.dylib' and version='0.1.2' it returns '.0.1.2.dylib'"""
+    Verbose = False
+    if Verbose:
+        print("_applelib_versioned_lib_suffix: suffix={:r}".format(suffix))
+        print("_applelib_versioned_lib_suffix: version={:r}".format(version))
+    if not version in suffix:
+        suffix = "." + version + suffix
+    if Verbose:
+        print("_applelib_versioned_lib_suffix: return suffix={:r}".format(suffix))
+    return suffix
+
+
+def _applelib_versioned_lib_soname(env, libnode, version, prefix, suffix, name_func):
+    """For libnode='/optional/dir/libfoo.X.Y.Z.dylib' it returns 'libfoo.X.dylib'"""
+    Verbose = False
+    if Verbose:
+        print("_applelib_versioned_lib_soname: version={!r}".format(version))
+    name = name_func(env, libnode, version, prefix, suffix)
+    if Verbose:
+        print("_applelib_versioned_lib_soname: name={!r}".format(name))
+    major = version.split('.')[0]
+    (libname,_suffix) = name.split('.')
+    soname = '.'.join([libname, major, _suffix])
+    if Verbose:
+        print("_applelib_versioned_lib_soname: soname={!r}".format(soname))
+    return soname
+
+def _applelib_versioned_shlib_soname(env, libnode, version, prefix, suffix):
+    return _applelib_versioned_lib_soname(env, libnode, version, prefix, suffix, link._versioned_shlib_name)
+
+
+# User programmatically describes how SHLIBVERSION maps to values for compat/current.
+_applelib_max_version_values = (65535, 255, 255)
+def _applelib_check_valid_version(version_string):
+    """
+    Check that the version # is valid.
+    X[.Y[.Z]]
+    where X 0-65535
+    where Y either not specified or 0-255
+    where Z either not specified or 0-255
+    :param version_string:
+    :return:
+    """
+    parts = version_string.split('.')
+    if len(parts) > 3:
+        return False, "Version string has too many periods [%s]"%version_string
+    if len(parts) <= 0:
+        return False, "Version string unspecified [%s]"%version_string
+
+    for (i, p) in enumerate(parts):
+        try:
+            p_i = int(p)
+        except ValueError as e:
+            return False, "Version component %s (from %s) is not a number"%(p, version_string)
+        if p_i < 0 or p_i > _applelib_max_version_values[i]:
+            return False, "Version component %s (from %s) is not valid value should be between 0 and %d"%(p, version_string, _applelib_max_version_values[i])
+
+    return True, ""
+
+
+def _applelib_currentVersionFromSoVersion(source, target, env, for_signature):
```

Comment:
So, my thought on the mailing list had been to not attempt to derive these from `SHLIBVERSION`, because it is just to hard to know what the right thing to do is.

My thought had been to levae it so the flags aren't added unless the user has explicitly set `APPLELINK_CURRENT_VERSION` and/or `APPLELINK__COMPATIBILITY_VERSION`.

If the user wants to automatically derive those values from `SHLIBVERSION`, they can write generators to do so, I think.

```
def currentVersionFromSoVersion(source, target, env, for_signature):
    return env['SOVERSION'].split(.)[1:]

def compatVersionFromSoVersion(source, target, env, for_signature):
    return env['SOVERSION'].split(.)[1] + ".0"

env['APPLELINK_CURRENT_VERSION'] = currentVersionFromSoVersion
env['APPLELINK_COMPATIBILITY_VERSION'] = compatVersionFromSoVersion

# Links as -Wl,current_version=2.3 -Wl,compat_version=2.0
env.SharedLibrary(..., SOVERSION=1.2.3, ...)
```

Reply:
That sounds fine to me. Thanks.